Mathlib/GroupTheory/PushoutI.lean

```diff
@@ -183,3 +190,486 @@ variable [∀ i, Group (G i)] [Group H] {φ : ∀ i, H →* G i}
+    equiv_fst_eq_mul_inv, mul_assoc, map_mul, map_inv, mul_smul, inv_smul_smul, summand_smul_def,
+    equivPair, rcons, Word.equivPair_symm, Word.rcons_eq_smul, Equiv.coe_fn_mk,
+    Word.equivPair_tail_eq_inv_smul, Equiv.coe_fn_symm_mk, smul_inv_smul]
+
+@[simp]
+theorem prod_summand_smul {i : ι} (g : G i) (w : NormalWord d) :
+    (g • w).prod = of i g * w.prod := by
+  simp only [prod, summand_smul_def', equivPair, rcons, Word.equivPair_symm,
+    Equiv.coe_fn_mk, Equiv.coe_fn_symm_mk, Word.equivPair_smul_same,
+    Word.equivPair_tail_eq_inv_smul, Word.rcons_eq_smul, ← of_apply_eq_base φ i,
+    MonoidHom.apply_ofInjective_symm, equiv_fst_eq_mul_inv, mul_assoc, map_mul, map_inv,
+    Word.prod_smul, ofCoprodI_of, inv_mul_cancel_left, mul_inv_cancel_left]
+
+@[simp]
+theorem prod_base_smul (h : H) (w : NormalWord d) :
+    (h • w).prod = base φ h * w.prod := by
+  simp only [base_smul_def', prod, map_mul, mul_assoc]
+
+@[simp]
+theorem prod_smul (g : PushoutI φ) (w : NormalWord d) :
+    (g • w).prod = g * w.prod := by
+  induction g using PushoutI.induction_on generalizing w with
+  | of i g => rw [of_smul_eq_smul, prod_summand_smul]
+  | base h => rw [base_smul_eq_smul, prod_base_smul]
+  | mul x y ihx ihy => rw [mul_smul, ihx, ihy, mul_assoc]
+
+@[simp]
+theorem prod_empty : (empty : NormalWord d).prod = 1 := by
+  simp [prod, empty]
+
+@[simp]
+theorem prod_cons {i} (g : G i) (w : NormalWord d) (hmw : w.fstIdx ≠ some i)
+    (hgr : g ∉ (φ i).range) : (cons g w hmw hgr).prod = of i g * w.prod := by
+  simp only [prod, cons, Word.prod, List.map, ← of_apply_eq_base φ i, equiv_fst_eq_mul_inv,
+    mul_assoc, MonoidHom.apply_ofInjective_symm, List.prod_cons, map_mul, map_inv,
+    ofCoprodI_of, inv_mul_cancel_left]
+
+theorem prod_smul_empty (w : NormalWord d) : w.prod • empty = w := by
+  induction w using consRecOn with
+  | h_empty => simp
+  | h_cons i g w _ _ _ _ ih =>
+    rw [prod_cons, mul_smul, ih, cons_eq_smul]
+  | h_base h w _ ih =>
+    rw [prod_smul, mul_smul, ih]
+
+/-- The equivalence between normal forms and elements of the pushout -/
+noncomputable def equiv : PushoutI φ ≃ NormalWord d :=
+  { toFun := fun g => g • .empty
+    invFun := fun w => w.prod
+    left_inv := fun g => by
+      simp only [prod_smul, prod_empty, mul_one]
+    right_inv := fun w => prod_smul_empty w }
+
+theorem prod_injective : Function.Injective (prod : NormalWord d → PushoutI φ) := by
+  letI := Classical.decEq ι
+  letI := fun i => Classical.decEq (G i)
+  classical exact equiv.symm.injective
+
+instance : FaithfulSMul (PushoutI φ) (NormalWord d) :=
+  ⟨fun h => by simpa using congr_arg prod (h empty)⟩
+
+instance (i : ι) : FaithfulSMul (G i) (NormalWord d) :=
+  ⟨by simp [summand_smul_def']⟩
+
+instance : FaithfulSMul H (NormalWord d) :=
+  ⟨by simp [base_smul_def']⟩
+
+end NormalWord
+
+open NormalWord
+
+theorem of_injective (hφ : ∀ i, Function.Injective (φ i)) (i : ι) :
```

Comment:
Could you add a docstring to this theorem and the next theorem, that would mention both the words `pushout` and `amalgamated product`?